Updates tutorial with entry on how to run Codex with its Markeplace (#765)
codex-storage/nim-codex: docs/Marketplace.md

We strongly suggest you to create a folder (e.g. `marketplace-tutorial`), and switch into it before beginning.

## 1. Set Up a Geth PoA Network

For this tutorial, we will use a simple [Proof-of-Authority](https://github.com/ethereum/EIPs/issues/225) network with geth. The first step is creating a _signer account_: an account which will be used by geth to sign the blocks in the network. Any block signed by a signer is accepted as valid.

### 1.1. Create a Signer Account

To create a signer account, run:

```bash
geth account new --datadir geth-data
```

The account generator will ask you to input a password, which you can leave blank. It will then print some information, including the account's public address:

```bash
INFO [03-22|12:58:05.637] Maximum peer count                       ETH=50 total=50
INFO [03-22|12:58:05.638] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
Your new account is locked with a password. Please give a password. Do not forget this password.
Password:
Repeat password:

Your new key was generated

Public address of the key:   0x93976895c4939d99837C8e0E1779787718EF8368
...
```

> In this example, the public address of the signer account is `0x93976895c4939d99837C8e0E1779787718EF8368`. Yours will print a different address; write it down.
> 

Reviewer: ```suggestion
In this example, the public address of the signer account is `0x93976895c4939d99837C8e0E1779787718EF8368`. Yours will print a different address. Save it for later usage.

Additionally set it as an environment variable for later usage:

```sh
export GETH_SIGNER_ADDR=0x93976895c4939d99837C8e0E1779787718EF8368
```

```

    }
  }
}
```

Note that the signer account address is embedded in two different places:
* inside of the `"extradata"` string, surrounded by zeroes and stripped of its `0x` prefix;
* as an entry key in the `alloc` session.

Make sure to replace that ID with the account ID that you wrote down in Step 1.1. Once that is done, you can initialize the network with:

```bash
geth init --datadir geth-data network.json
```

### 1.3. Start your PoA Node

We are now ready to start our $1$-node, private blockchain. To launch the signer node, open a separate terminal on the same working directory and run:

```bash
geth\
  --datadir geth-data\
  --networkid 12345\
  --unlock <signer-account-address>\
  --nat extip:127.0.0.1\
  --netrestrict 127.0.0.0/24\
  --mine\
  --miner.etherbase <signer-account-address>\
  --http\
  --allow-insecure-unlock
```

Note that, once again, the signer account created in Step 1.1 appears both in `--unlock` and `--allow-insecure-unlock`: do not forget to insert it there. Geth will prompt you to insert the account's password as it starts up. Once you do that, it should be able to start up and begin "mining" blocks.

## 2. Set Up The Marketplace

> Setting up the marketplace entails: _i)_ deploying the Codex Marketplace contracts to our private blockchain, and _ii)_ provisioning accounts we will later use to buying and selling storage in the marketplace with the required token balances.

Reviewer: Numbered bullets here are probably a bit easier to follow (for me at least). Also `buying/selling` don't match the rest of the sentence:

```suggestion
Setting up the Codex marketplace entails:

1. Deploying the Codex Marketplace contracts to our private blockchain
2. Setup Ethereum accounts we will use to buy and sell storage in the Codex marketplace 
3. Provisioning those accounts with the required token balances

```

git clone https://github.com/codex-storage/codex-contracts-eth
cd codex-contracts-eth
npm install
```
You now must **wait until $256$ blocks are mined in your PoA network**, or deploy will fail. This should take about $4$ minutes and $30$ seconds. You can check which block height you are currently at by running:

```bash
geth attach --exec web3.eth.blockNumber ./geth-data/geth.ipc
```

once that gets past $256$, you are ready to go. To deploy contracts, run:

```bash
export DISTTEST_NETWORK_URL=http://localhost:8545 # bootstrap node
npx hardhat --network codexdisttestnetwork deploy
```

If the command completes successfully, you are ready to prepare the accounts.

### 2.2. Generate the Required Accounts

We will run $2$ Codex nodes: a **storage provider**, which will sell storage on the network, and a **client**, which will buy and use such storage; we therefore need two valid Ethereum accounts. We could create random accounts by using one of the many  tools available to that end but, since this is a tutorial running on a local private network, we will simply provide you with two pre-made accounts along with their private keys which you can copy and paste instead:

**Storage:**

> ```text
> address: 0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37
> private key: 0x06c7ac11d4ee1d0ccb53811b71802fa92d40a5a174afad9f2cb44f93498322c3

Reviewer: ```suggestion
```sh
ETH_STORAGE_ADDR=0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37
ETH_STORAGE_PK=0x06c7ac11d4ee1d0ccb53811b71802fa92d40a5a174afad9f2cb44f93498322c3
```

```
**Client:**

> ```text
> address: 0x9F0C62Fe60b22301751d6cDe1175526b9280b965
> private key: 0x5538ec03c956cb9d0bee02a25b600b0225f1347da4071d0fd70c521fdc63c2fc

Reviewer: ```suggestion
```sh
ETH_CLIENT_ADDR=0x9F0C62Fe60b22301751d6cDe1175526b9280b965
ETH_CLIENT_PK=0x5538ec03c956cb9d0bee02a25b600b0225f1347da4071d0fd70c521fdc63c2fc
```

```

### 2.3. Provision Accounts with Tokens

We now need to transfer some ETH to each of the accounts, as well as provide them with some Codex tokens for the storage node to use as collateral and for the client node to buy actual storage.

Although the process is not particularly complicated, I suggest you use [the script we prepared](https://github.com/gmega/local-codex-bare/blob/main/scripts/mint-tokens.js) for that. This script, essentially:

1. reads the Marketplace contract address and its ABI from the deployment data;
2. transfers $1$ ETH from the signer account to a target account if the target account has no ETH balance;
3. mints $n$ Codex tokens and adds it into the target account's balance.

> To use the script, just copy it locally into a file named `mint-tokens.js`, and run:
> 
> ```bash
> # Installs Web3-js

Reviewer: ```suggestion
To use the script, just download it into a local file named `mint-tokens.js` or use curl:

```bash
curl https://github.com/gmega/local-codex-bare/blob/main/scripts/mint-tokens.js -o mint-token.js
```

Then run the following to mine the tokens:
```bash
# Installs Web3-js
```

npm install web3
# Provides tokens to the storage account.

> node ./mint-tokens.js <signer account address> 0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37 10000000000
> # Provides tokens to the client account.
> node ./mint-tokens.js <signer account address> 0x9F0C62Fe60b22301751d6cDe1175526b9280b965 10000000000

Reviewer: ```suggestion
node ./mint-tokens.js $GETH_SIGNER_ADDR 0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37 10000000000
# Provides tokens to the client account.
node ./mint-tokens.js $GETH_SIGNER_ADDR 0x9F0C62Fe60b22301751d6cDe1175526b9280b965 10000000000
```

> # Provides tokens to the storage account.
> node ./mint-tokens.js <signer account address> 0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37 10000000000
> # Provides tokens to the client account.
> node ./mint-tokens.js <signer account address> 0x9F0C62Fe60b22301751d6cDe1175526b9280b965 10000000000

Reviewer: ```suggestion
# set the contract file location
export CONTRACT_DEPLOY_FULL="codex-contracts-eth/deployments/codexdisttestnetwork"

# Provides tokens to the storage account.
node ./mint-tokens.js $GETH_SIGNER_ADDR 0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37 10000000000

# Provides tokens to the client account.
node ./mint-tokens.js $GETH_SIGNER_ADDR 0x9F0C62Fe60b22301751d6cDe1175526b9280b965 10000000000
```

```

> Don't forget to replace `<signer account address>` with the address of the signer account you created in Step 1.1.

Reviewer: ```suggestion
Don't forget to set the variable `GETH_SIGNER_ADDR` with the address of the signer account you created in Step 1.1. You can manually replace `$GETH_SIGNER_ADDR` with the address as well.
```

## 3. Run Codex

With accounts and geth in place, we can now start the Codex nodes.

### 3.1. Storage Node

The storage node will be the one storing data and submitting the proofs of storage to the chain. To do that, it needs access to:

1. the address of the Marketplace contract that has been deployed to the local geth node in [Step 2.1](#21-deploy-the-codex-marketplace-contracts);
2. the sample ceremony files which are shipped in the Codex contracts repo.

Recall you have clone the `codex-contracts-eth` repository in Step 2.1. All of the required files are in there.

**Address of the Marketplace Contract.** The contract address can be found inside of the file `codex-contracts-eth/deployments/codexdisttestnetwork/Marketplace.json`:

```bash
> grep '"address":' Marketplace.json
  "address": "0x8891732D890f5A7B7181fBc70F7482DE28a7B60f",
```

> 

Reviewer: Then run the following with the correct market place address:

```sh
export MARKETPLACE_ADDRESS=0x8891732D890f5A7B7181fBc70F7482DE28a7B60f
```

**Prover ceremony files.** The ceremony files are under the `codex-contracts-eth/verifier/networks/codexdisttestnetwork` subdirectory. There are three of them: `proof_main.r1cs`, `proof_main.zkey`, and `prooof_main.wasm`. We will need all of them to start the Codex storage node.

> 

Reviewer: ```suggestion

```sh
export CONTRACT_DEPLOY_FULL="codex-contracts-eth/deployments/codexdisttestnetwork"
export CONTRACT_VERIFIER=codex-contracts-eth/verifier/networks/codexdisttestnetwork/
```

Reviewer: I don't know how to quote the backticks inside the backticks. 😢

**Starting the storage node.** Let:

* `PROVER_ASSETS` contain the directory where the prover ceremony files are located. **This must be an absolute path**;
* `CODEX_BINARY` contain the location of your Codex binary;
* `MARKETPLACE_ADDRESS` contain the address of the Marketplace contract (obtained above).

To launch the storage node, run:

```bash
${CODEX_BINARY}\
  --data-dir=./codex-storage\
  --listen-addrs=/ip4/0.0.0.0/tcp/8080\
  --api-port=8000\
  --disc-port=8090\
  persistence\
  --eth-provider=http://localhost:8545\

>   --eth-private-key=<(echo -n "0x06c7ac11d4ee1d0ccb53811b71802fa92d40a5a174afad9f2cb44f93498322c3")\

Reviewer: ```suggestion
  --eth-private-key=./storage.pkey\
```

  --marketplace-address=${MARKETPLACE_ADDRESS}\
  --validator\
  --validator-max-slots=1000\
  prover\
  --circom-r1cs=${PROVER_ASSETS}/proof_main.r1cs\
  --circom-wasm=${PROVER_ASSETS}/proof_main.wasm\
  --circom-zkey=${PROVER_ASSETS}/proof_main.zkey
```
replacing each `${VALUE}` variable by their respective contents. We then extract the Signed Peer Record (SPR) of the storage node so we can bootstrap the client node with it. To get the SPR, issue the following call:

```bash

> curl 'http://localhost:8080/api/codex/v1/debug/info'

Reviewer: ```suggestion
curl 'http://localhost:8000/api/codex/v1/debug/info'
```

```

This will print a long JSON string:

```json
{"id":"16Uiu2HAm4yoghEaNy1VPLumvXwrFSEenzG1zSy3EcTFYLE2mynjp","addrs":["/ip4/0.0.0.0/tcp/8070"],"repo":"/data","spr":"spr:CiUIAhIhAo30ewLoBnzU5COMicomRT6UAfdmDJbdwQRnba0cvHnnEgIDARo8CicAJQgCEiECjfR7AugGfNTkI4yJyiZFPpQB92YMlt3BBGdtrRy8eecQqrWcsAYaCwoJBLNdyciRAh-aKkYwRAIgFnH5tSmOlxrSnOwyOHjnjBVSrEXZOeosH1i9gzoDI6UCIEnCI82Bb4iATMKKuaZCfVNlEOmwnfAG-6z0RuJJc-8s",...
```
The part we care about is the SPR (the string after `"spr":`). Write that string down.

**Starting the client node.** The client node is started similarly except that:

* we need to pass the SPR of the storage node so it can form a network with it;
* since it does not run any proofs, it does not require any ceremony files.

```bash
${CODEX_BINARY}\
  --data-dir=./codex-client\
  --listen-addrs=/ip4/0.0.0.0/tcp/8081\
  --api-port=8001\
  --disc-port=8091\
  --bootstrap-node=<storage-node-spr>
  persistence\
  --eth-provider=http://localhost:8545\

>   --eth-private-key=<(echo -n "0x5538ec03c956cb9d0bee02a25b600b0225f1347da4071d0fd70c521fdc63c2fc")\

Reviewer: ```suggestion
  --eth-private-key=./client.pkey\
```

  --marketplace-address=${MARKETPLACE_ADDRESS}
```

## 4. Buy and Sell Storage on the Marketplace

Any storage negotiation has two sides: a buyer and a seller. Before we can actually request storage, therefore, we must first put some of it for sale.

### 4.1 Sell Storage

> The following request will cause the storage node to put out $50\text{MB}$ of storage for sale for $1$ hour, at a price of $1$ Codex token per byte per second, while expressing that it's willing to take at most a $1000$ Codex token penalty for missing a storage proof.

Author: Thanks @AuHau, I've rephrased and added a footnote.

```bash
curl 'http://localhost:8000/api/codex/v1/sales/availability' \
  --header 'Content-Type: application/json' \
  --data '{
  "totalSize": "50000000",
  "duration": "3600",
  "minPrice": "1",
  "maxCollateral": "1000"
}'
```

this should return a response with an id a string (e.g. `"id": "0x552ef12a2ee64ca22b237335c7e1df884df36d22bfd6506b356936bc718565d4"`) which identifies this storage offer. To check the current storage offers for this node, you can issue:

```bash
curl 'http://localhost:8000/api/codex/v1/sales/availability'
```

this should print a list of offers, with the one you just created figuring among them.

## 4.2. Buy Storage

Before we can buy storage, we must have some actual data to request storage for. Start by uploading a small file to your client node. On Linux you could, for instance, use `dd` to generate a $100KB$ file:

```bash
dd if=/dev/urandom of=./data.bin bs=100K count=1
```

but any small file will do. Assuming your file is named `data.bin`, you can upload it with:

```bash

> curl "http://localhost:8001/api/codex/v1/data" --data-bin @data1.bin

Reviewer: ```suggestion
curl "http://localhost:8001/api/codex/v1/data" --data-bin ./data1.bin
```

```

Once the upload completes, you should see a CID (e.g. `zDvZRwzm2mK7tvDzKScRLapqGdgNTLyyEBvx1TQY37J2CdWdS6Sj`) for the file printed to the terminal. Use that CID in the purchase request:

```bash

> curl "http://localhost:8081/api/codex/v1/storage/request/<upload CID>"

Reviewer: ```suggestion
curl "http://localhost:8001/api/codex/v1/storage/request/<upload CID>"
```

Reviewer: I'm getting an error I think you ran into previously: `Unable to encode manifest, not enough blocks, ecK = 2, blocksCount = 1%`

  --header 'Content-Type: application/json' \
  --data '{
	"duration": "1200",
	"reward": "1",
	"proofProbability": "3",
	"expiry": "1711992852",
	"nodes": 3,
	"tolerance": 1,
	"collateral": "1000"
}'
```

The parameters under `--data` say that:

1. we want to purchase storage for our file for $20$ minutes (`"duration": "1200"`);
2. we are willing to pay up to $1$ token per byte, per second (`"reward": "1"`);

> 3. we demand `1000` tokens in collateral for this storage request. Without getting into too much detail, this means we will be compensated with `1000` tokens for each partition in our file that storage nodes fail to provide timely proofs for;

Reviewer 2: When slashing nobody is "compensated", instead the tokens are mainly burned. There is also small portion of the tokens kept as an incentive reward for the future repair mechanism in order to compensate the nodes for the extra work they will need to do for the repair.

With specifying the collateral the user can indicate "how much valuable" is the data for him. Where most probably for higher collateral (and hence for the nodes having higher stake = motivation in the hosting of the content), the user will have to pay a premium price.

Author: Thanks, did some slight rephrasing which together with the footnote above should make it clear enough (I hope).

### 1.1. Create a Signer Account

To create a signer account, run:

```bash
geth account new --datadir geth-data
```

The account generator will ask you to input a password, which you can leave blank. It will then print some information, including the account's public address:

```bash
INFO [03-22|12:58:05.637] Maximum peer count                       ETH=50 total=50
INFO [03-22|12:58:05.638] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
Your new account is locked with a password. Please give a password. Do not forget this password.
Password:
Repeat password:

Your new key was generated

Public address of the key:   0x93976895c4939d99837C8e0E1779787718EF8368
...
```

In this example, the public address of the signer account is `0x93976895c4939d99837C8e0E1779787718EF8368`. Yours will print a different address. Save it for later usage.

Next set an environment variable for later usage:

```sh

> export GETH_SIGNER_ADDR=0x0000000000000000000000000000000000000000

Reviewer: ```suggestion
export GETH_SIGNER_ADDR="0x0000000000000000000000000000000000000000"
```

```

If you get a message like `Usage: mint-tokens.js <token-hardhat-deploy-json> <signer-account> <receiver-account> <token-ammount>` then you need to ensure you have 

## 3. Run Codex

With accounts and geth in place, we can now start the Codex nodes.

### 3.1. Storage Node

The storage node will be the one storing data and submitting the proofs of storage to the chain. To do that, it needs access to:

1. the address of the Marketplace contract that has been deployed to the local geth node in [Step 2.1](#21-deploy-the-codex-marketplace-contracts);
2. the sample ceremony files which are shipped in the Codex contracts repo.

Recall you have clone the `codex-contracts-eth` repository in Step 2.1. All of the required files are in there.

**Address of the Marketplace Contract.** The contract address can be found inside of the file `codex-contracts-eth/deployments/codexdisttestnetwork/Marketplace.json`:

```bash
grep '"address":' ${CONTRACT_DEPLOY_FULL}/Marketplace.json
```

which should print something like:
```sh
  "address": "0x8891732D890f5A7B7181fBc70F7482DE28a7B60f",
```

Then run the following with the correct market place address:
```sh

> export MARKETPLACE_ADDRESS=0x0000000000000000000000000000000000000000

Reviewer: ```suggestion
export MARKETPLACE_ADDRESS="0x0000000000000000000000000000000000000000"
```
